README.md
Sample App: Marko + Express
======================================

This sample app illustrates how to integrate Marko with a very basic Express app. For this sample app, we use the streaming API to stream the output of the template rendering to the HTTP response stream. In addition, this sample application illustrates how to create custom tags that can be embedded into your templates.

# Installation

```
git clone https://github.com/marko-js-samples/marko-express.git
cd marko-express
npm install
node server.js
```

Navigate to [http://localhost:8080/](http://localhost:8080/) to see your server in action!

# Project Structure

```
.
├── components - Directory containing custom tag implementations
│   ├── app-button - Custom tag for rendering a Bootstrap-styled button
│   │   ├── renderer.js
│   │   └── template.marko
│   ├── app-header - Custom tag for rendering the page header (template only)
│   │   └── template.marko
│   └── app-hello - Custom tag rendering a simple message (JavaScript renderer only)
│       └── renderer.js
├── footer.marko - An include target
├── index.marko - The page template
├── package.json - npm metadata
├── marko-taglib.json - Marko taglib used to discover custom tags
├── server.js - JavaScript entry point for this application
└── static - Folder containing static JavaScript and CSS files
    ├── reset.css
    └── style.css
```

# Details

## Page Rendering

This application registers a single "/" route that renders out the main index page using the `index.marko` template.

The template is loaded using the following code:

```javascript
var indexTemplate = require('./index.marko');
```

Later, in the route handler, the template is rendered writable HTTP response stream as shown in the code below:

```javascript
indexTemplate.stream({
        name: 'Frank',
        count: 30,
        colors: ['red', 'green', 'blue']
    }, res);
```

As with all Node.js streams, when piping to a target stream, the target stream is ended when the source stream is done producing data. Since we are piping to an HTTP response stream, the HTTP response stream is ended and the connection is closed.

## Custom Tags

This sample app includes a `marko-taglib.json` in the root that is used to register the custom tags. This taglib file is automatically discovered by searching up the directory tree from a template's location on disk. For this sample, the `marko-taglib.json` is the following:

```json
{
    "tags-dir": "./components"
}
```

This simple taglib tells the Marko compiler to scan the `components` directory to discover custom tags. In this sample app, the following three custom tags will be discovered and registered:

1. `<app-button/>`
2. `<app-header/>`
3. `<app-hello/>`

The `<app-button/>` custom tag uses `marko-tag.json` file to declare which attributes are allowed. This tag definition is shown below:

```javascript
{
    "@label": "string",
    "@href": "string",
    "@variant": "string", // primary | info | success | warning | danger | inverse
    "@size": "string", // large | small | mini
    "@toggle": "boolean",
    "@toggled": "boolean",
    "@dropdown": "boolean",
    "@*": "string"
}
```

## Includes

This sample app also illustrates how to use the `<include>` tag to include another template. In `index.marko` the following code is used to include the footer template:

```html
<include template="./footer.marko"/>
```

_NOTE: Custom tags are often a better choice over using template includes. In this sample app, the header is a custom tag and the footer is an include so that you can see the difference._
"# mob_mfk_marko" 


history.txt
   12  cd ~
   13  curl -sL https://deb.nodesource.com/setup_10.x | sudo -E bash -
   14  sudo apt-get install -y nodejs
   15  sudo apt install node
   16  node -v
   17  sudo apt-get install -y build-essential
   18  npm -v
   19  node -v
   20  node -v
   21  npm -v
   22  adv
   23  avb
   24  avd
   25  adb devices
   26  adb devices
   27  adb devices
   28  npm start
   29  sudo npm start
   30  npm install --global expo-cli
   31  sudo npm install --global expo-cli
   32  cd $HOME
   33  ls -al
   34  cd mnt
   35  cd mnt/
   36  cd /mnt/
   37  ls -al
   38  cd ~
   39  nano .bashrc
   40  ls -al
   41  nano .bash_profile
   42  source .bash_profile
   43  rm -r node_modules
   44  ls -al
   45  npm start
   46  фви
   47  adb
   48  adb version
   49  nano ~/.bashrc
   50  npm install prerender
   51  npm audit fix
   52  node server.js start
   53  node server.js start
   54  cd ..//
   55  export PRERENDER_SERVICE_URL=http://localhost:8081
   56  git clone https://github.com/prerender/prerender.git
   57  cd prerender
   58  npm install
   59  npm audit fix
   60  node server.js
   61  cls
   62  wget -q -O - https://dl-ssl.google.com/linux/linux_signing_key.pub | sudo apt-key add -
   63  echo 'deb [arch=amd64] http://dl.google.com/linux/chrome/deb/ stable main' | sudo tee /etc/apt/sources.list.d/google-chrome.list
   64  sudo apt-get update 
   65  sudo apt-get install google-chrome-stable
   66  node server.js
   67  whereis google-chrome
   68  node server.js
   69  node server.js
   70  whereis google-chrome
   71  /usr/bin/google-chrome
   72  sudo apt install gksu
   73  sudo apt install gksu
   74  sudo /usr/bin/google-chrome
   75  sudo /usr/bin/google-chrome --no-sandbox
   76  sudo /usr/bin/google-chrome --no-sandbox
   77  exit
   78  sudo /usr/bin/google-chrome
   79  google-chrome
   80  sudo rm ~/.config/google-chrome/*
   81  sudo rm -rf ~/.config/google-chrome/*
   82  node server.js
   83  node server.js
   84  node server.js
   85  node server.js
   86  node server.js
   87  expo start
   88  npm install -g expo-cli
   89  history >> history.txt
   90  sudo chown -R victor: /usr/lib/node_modules
   91  123
   92  npm install -g expo-cli
   93  sudo chown -R victor: /usr/bin/expo/
   94  sudo chown -R victor: /usr/bin/expo
   95  sudo chown -R victor: /usr/bin/expo/
   96  npm install -g expo-cli
   97  sudo chown -R victor: /usr/bin/expo-cli
   98  sudo chown -R victor: /usr/bin/expo-cli/
   99  npm install -g expo-cli
  100  sudo npm install -g expo-cli
  101  sudo chown -R victor: /home/victor/.npm
  102  sudo chown -R victor: /usr/lib/node_modules/expo-cli/node_modules
  103  sudo npm install -g expo-cli
  104  sudo chown -R victor: /home/victor/.npm
  105  123
  106  sudo chown -R victor: /usr/lib/node_modules/expo-cli/node_modules/
  107  sudo npm install -g expo-cli
  108  expo start
  109  adb devices
  110  expo start --localhost --android
  111  sudo npm install react-native-elements
  112  sudo npm install react-native-elements
  113  npm audit fix
  114  expo start --localhost --android
  115  watchman watch-del /mnt/d/node/notifications ; watchman watch-project /mnt/d/node/notifications
  116  expo start --localhost --android
  117  sudo npm install react-navigation
  118  sudo npm install react-navigation
  119  sudo chown -R victor: /mnt/d/node/notifications/node_modules/
  120  sudo npm install react-navigation
  121  sudo chown -R victor: '/mnt/d/node/notifications/
  122  '
  123  sudo chown -R victor: /mnt/d/node/notifications/
  124  sudo npm install react-navigation
  125  npm install -s react-navigation
  126  expo start --localhost --android
  127  watchman watch-del /mnt/d/node/notifications ; watchman watch-project /mnt/d/node/notifications
  128  expo start --localhost --android
  129  watchman watch-del-all
  130  rm -rf node_modules
  131  npm install
  132  npm audit fix
  133  expo start --localhost --android
  134  кь -к
  135  rm -rf /tmp/metro-bundler-cache-*
  136  rm -rf /tmp/metro-bundler-cache-*
  137  expo r -c
  138  adb
  139  adb devices
  140  npm start -- --reset-cache
  141  expo start --reset-cache
  142  rm -rf /tmp/metro-bundler-cache-*
  143  expo start --localhost --android
  144  cd ..//
  145  expo init rn1
  146  cd rn1
  147  expo start --localhost --androif
  148  expo start --localhost --android
  149  npm install react-navigation
  150  expo start --localhost --android
  151  npm install
  152  npm install
  153  sudo chown -R victor: '/mnt/d/node/rn1/
  154  sudo chown -R victor: /mnt/d/node/rn1/
  155  npm install
  156  cd ..//
  157  expo init rn2
  158  expo init rn3
  159  expo init rn4
  160  cd rn4
  161  npm init
  162  npm install
  163  npm audit fix
  164  yarn add react-navigation
  165  yarn add react-native-elements
  166  mkdir components
  167  touch components/HomeScreen.js
  168  touch components/DetailsScreen.js
  169  yarn start --localhost --android
  170  yarn add react-native-gesture-handler
  171  yarn start --localhost --android
  172  expo install react-native-gesture-handler react-native-reanimated
  173  куфсе-тфешму -м
  174  react-native -v
  175  react-native -v
  176  reactnative -v
  177  react-native
  178  yarn start --localhost --android
  179  node service
  180  node server
  181  npm install
  182  npm audit fix
  183  adb devices
  184  adb devices
  185  adb devices
  186  yarn start --localhost --android
  187  yarn start --localhost --android
  188  yarn start --localhost --android
  189  node server
  190  node server
  191  node server
  192  node server
  193  node server
  194  node server.js
  195  forever start node server.js
  196  forever start server.js
  197  npm install forever -g
  198  sudo npm install forever -g
  199  npm install -g npm
  200  sudo npm install -g npm
  201  forever start server.js
  202  forever -l
  203  forever list
  204  nano /home/victor/.forever/auit.log
  205  nano /home/victor/.forever/auit.log

src/jslib/enother.js
const cache = require('memory-cache');

exports.MD5 = (d) => {result = M(V(Y(X(d),8*d.length)));return result.toLowerCase()};function M(d){for(var _,m="0123456789ABCDEF",f="",r=0;r<d.length;r++)_=d.charCodeAt(r),f+=m.charAt(_>>>4&15)+m.charAt(15&_);return f}function X(d){for(var _=Array(d.length>>2),m=0;m<_.length;m++)_[m]=0;for(m=0;m<8*d.length;m+=8)_[m>>5]|=(255&d.charCodeAt(m/8))<<m%32;return _}function V(d){for(var _="",m=0;m<32*d.length;m+=8)_+=String.fromCharCode(d[m>>5]>>>m%32&255);return _}function Y(d,_){d[_>>5]|=128<<_%32,d[14+(_+64>>>9<<4)]=_;for(var m=1732584193,f=-271733879,r=-1732584194,i=271733878,n=0;n<d.length;n+=16){var h=m,t=f,g=r,e=i;f=md5_ii(f=md5_ii(f=md5_ii(f=md5_ii(f=md5_hh(f=md5_hh(f=md5_hh(f=md5_hh(f=md5_gg(f=md5_gg(f=md5_gg(f=md5_gg(f=md5_ff(f=md5_ff(f=md5_ff(f=md5_ff(f,r=md5_ff(r,i=md5_ff(i,m=md5_ff(m,f,r,i,d[n+0],7,-680876936),f,r,d[n+1],12,-389564586),m,f,d[n+2],17,606105819),i,m,d[n+3],22,-1044525330),r=md5_ff(r,i=md5_ff(i,m=md5_ff(m,f,r,i,d[n+4],7,-176418897),f,r,d[n+5],12,1200080426),m,f,d[n+6],17,-1473231341),i,m,d[n+7],22,-45705983),r=md5_ff(r,i=md5_ff(i,m=md5_ff(m,f,r,i,d[n+8],7,1770035416),f,r,d[n+9],12,-1958414417),m,f,d[n+10],17,-42063),i,m,d[n+11],22,-1990404162),r=md5_ff(r,i=md5_ff(i,m=md5_ff(m,f,r,i,d[n+12],7,1804603682),f,r,d[n+13],12,-40341101),m,f,d[n+14],17,-1502002290),i,m,d[n+15],22,1236535329),r=md5_gg(r,i=md5_gg(i,m=md5_gg(m,f,r,i,d[n+1],5,-165796510),f,r,d[n+6],9,-1069501632),m,f,d[n+11],14,643717713),i,m,d[n+0],20,-373897302),r=md5_gg(r,i=md5_gg(i,m=md5_gg(m,f,r,i,d[n+5],5,-701558691),f,r,d[n+10],9,38016083),m,f,d[n+15],14,-660478335),i,m,d[n+4],20,-405537848),r=md5_gg(r,i=md5_gg(i,m=md5_gg(m,f,r,i,d[n+9],5,568446438),f,r,d[n+14],9,-1019803690),m,f,d[n+3],14,-187363961),i,m,d[n+8],20,1163531501),r=md5_gg(r,i=md5_gg(i,m=md5_gg(m,f,r,i,d[n+13],5,-1444681467),f,r,d[n+2],9,-51403784),m,f,d[n+7],14,1735328473),i,m,d[n+12],20,-1926607734),r=md5_hh(r,i=md5_hh(i,m=md5_hh(m,f,r,i,d[n+5],4,-378558),f,r,d[n+8],11,-2022574463),m,f,d[n+11],16,1839030562),i,m,d[n+14],23,-35309556),r=md5_hh(r,i=md5_hh(i,m=md5_hh(m,f,r,i,d[n+1],4,-1530992060),f,r,d[n+4],11,1272893353),m,f,d[n+7],16,-155497632),i,m,d[n+10],23,-1094730640),r=md5_hh(r,i=md5_hh(i,m=md5_hh(m,f,r,i,d[n+13],4,681279174),f,r,d[n+0],11,-358537222),m,f,d[n+3],16,-722521979),i,m,d[n+6],23,76029189),r=md5_hh(r,i=md5_hh(i,m=md5_hh(m,f,r,i,d[n+9],4,-640364487),f,r,d[n+12],11,-421815835),m,f,d[n+15],16,530742520),i,m,d[n+2],23,-995338651),r=md5_ii(r,i=md5_ii(i,m=md5_ii(m,f,r,i,d[n+0],6,-198630844),f,r,d[n+7],10,1126891415),m,f,d[n+14],15,-1416354905),i,m,d[n+5],21,-57434055),r=md5_ii(r,i=md5_ii(i,m=md5_ii(m,f,r,i,d[n+12],6,1700485571),f,r,d[n+3],10,-1894986606),m,f,d[n+10],15,-1051523),i,m,d[n+1],21,-2054922799),r=md5_ii(r,i=md5_ii(i,m=md5_ii(m,f,r,i,d[n+8],6,1873313359),f,r,d[n+15],10,-30611744),m,f,d[n+6],15,-1560198380),i,m,d[n+13],21,1309151649),r=md5_ii(r,i=md5_ii(i,m=md5_ii(m,f,r,i,d[n+4],6,-145523070),f,r,d[n+11],10,-1120210379),m,f,d[n+2],15,718787259),i,m,d[n+9],21,-343485551),m=safe_add(m,h),f=safe_add(f,t),r=safe_add(r,g),i=safe_add(i,e)}return Array(m,f,r,i)}function md5_cmn(d,_,m,f,r,i){return safe_add(bit_rol(safe_add(safe_add(_,d),safe_add(f,i)),r),m)}function md5_ff(d,_,m,f,r,i,n){return md5_cmn(_&m|~_&f,d,_,r,i,n)}function md5_gg(d,_,m,f,r,i,n){return md5_cmn(_&f|m&~f,d,_,r,i,n)}function md5_hh(d,_,m,f,r,i,n){return md5_cmn(_^m^f,d,_,r,i,n)}function md5_ii(d,_,m,f,r,i,n){return md5_cmn(m^(_|~f),d,_,r,i,n)}function safe_add(d,_){var m=(65535&d)+(65535&_);return(d>>16)+(_>>16)+(m>>16)<<16|65535&m}function bit_rol(d,_){return d<<_|d>>>32-_}

exports.genGuid = () => {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
    const r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
    return v.toString(16);
  });
}

exports.storeSecretKey = ( timeMs = 2000 ) => {

  const MD5guid = MD5(MD5(genGuid()));
  const guid = genGuid() + genGuid();

  cache.put( MD5guid, guid, timeMs, function(key, value) {
    //console.log(`${MD5guid} did  ${value}` );
  });

  return { key: MD5guid, value: guid }

}

exports.secretPairOk = ( params ) => { return params.key ? (cache.get( params.key ) === params.value) : false }
